File: ScanbotSDKDemo/GenericDocumentsDemo/GenericDocumentViewController.m
//
//  GenericDocumentViewController.m
//  SBSDK Internal Demo
//
//  Created by Sebastian Husche on 22.05.20.
//  Copyright © 2020 doo GmbH. All rights reserved.
//

@import ScanbotSDK;
#import "GenericDocumentViewController.h"
#import "GenericDocumentResultViewController.h"


@interface GenericDocumentViewController () <SBSDKGenericDocumentRecognizerViewControllerDelegate, UINavigationControllerDelegate,
UIImagePickerControllerDelegate>

@property (nonatomic, strong) SBSDKGenericDocumentRecognizerViewController *scannerViewController;
@property (nonatomic, strong) UIActivityIndicatorView *indicator;

@end

@implementation GenericDocumentViewController

- (NSArray <SBSDKGenericDocumentRootType *>*)documentTypes {
    return [SBSDKGenericDocumentRootType allDocumentTypes];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.scannerViewController =
    [[SBSDKGenericDocumentRecognizerViewController alloc] initWithParentViewController:self
                                                                            parentView:self.view
                                                                 acceptedDocumentTypes:[self documentTypes]
                                                                              delegate:self];
    
    self.indicator = [[UIActivityIndicatorView alloc]initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
    self.indicator.hidden = YES;
    [self.view addSubview:self.indicator];
    
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    self.indicator.center = CGPointMake(CGRectGetMidX(self.view.bounds), CGRectGetMidY(self.view.bounds));
}

- (void)documentRecognizerViewController:(nonnull SBSDKGenericDocumentRecognizerViewController *)controller
                    didRecognizeDocument:(nonnull SBSDKGenericDocument *)document
                                 onImage:(nonnull UIImage *)image {
    
    self.indicator.hidden = YES;
    [self.indicator stopAnimating];

    [self displayResult:document withSourceImage:image];

}

- (void)displayResult:(SBSDKGenericDocument *)document withSourceImage:(UIImage *)sourceImage {
    
    if (document != nil && self.navigationController.topViewController == self) {
        GenericDocumentResultViewController *resultsVC = [GenericDocumentResultViewController makeWithResult:document
                                                                                             sourceImage:sourceImage];

        [self.navigationController pushViewController:resultsVC animated:YES];
    }
}

- (IBAction)selectImageButtonTapped:(id)sender {
    UIImagePickerController *picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    [self presentViewController:picker animated:YES completion:nil];
}

#pragma mark - UIImagePickerController delegate

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [picker.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    UIImage *image = (UIImage *)info[UIImagePickerControllerOriginalImage];
    [self dismissViewControllerAnimated:YES completion:^{
        SBSDKGenericDocumentRecognizer *recognizer =
        [[SBSDKGenericDocumentRecognizer alloc] initWithAcceptedDocumentTypes:[self documentTypes]];
        
        SBSDKGenericDocument *document = [recognizer recognizeDocumentOnImage:image];
        [self displayResult:document withSourceImage:image];
    }];
}

@end

File: ScanbotSDKDemo/GenericDocumentsDemo/GenericDocumentViewController.swift
//
//  GenericDocumentViewController.swift
//  ScanbotSDKDemo
//
//  Created by Danil Voitenko on 23.11.21.
//  Copyright © 2021 doo GmbH. All rights reserved.
//

import UIKit
import ScanbotSDK

class GenericDocumentViewController: UIViewController {
    var scannerViewController: SBSDKGenericDocumentRecognizerViewController?
    var indicator: UIActivityIndicatorView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        scannerViewController = SBSDKGenericDocumentRecognizerViewController(parentViewController: self,
                                                                             parentView: self.view,
                                                                             acceptedDocumentTypes: documentTypes(),
                                                                             delegate: self)
        
        indicator = UIActivityIndicatorView(style: .whiteLarge)
        indicator?.hidesWhenStopped = true
        view.addSubview(indicator!)
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        indicator?.center = CGPoint(x: view.bounds.midX, y: view.bounds.midY)
    }
    
    private func documentTypes() -> [SBSDKGenericDocumentRootType] {
        return SBSDKGenericDocumentRootType.allDocumentTypes()
    }
    
    private func display(document: SBSDKGenericDocument, with sourceImage: UIImage) {
        if navigationController?.topViewController == self {
            let resultsVC = GenericDocumentResultViewController.make(with: document, sourceImage: sourceImage)
            navigationController?.pushViewController(resultsVC, animated: true)
        }
    }
    
    @IBAction private func selectImageButtonDidPress(_ sender: Any) {
        let picker = UIImagePickerController()
        picker.delegate = self
        present(picker, animated: true, completion: nil)
    }
}

extension GenericDocumentViewController: SBSDKGenericDocumentRecognizerViewControllerDelegate {
    func documentRecognizerViewController(_ controller: SBSDKGenericDocumentRecognizerViewController,
                                          didRecognizeResult result: SBSDKGenericDocumentRecognitionResult,
                                          on image: UIImage) {
        if result.status == .success {
            indicator?.stopAnimating()
        }
        if let document = result.document, let sourceImage = result.croppedImage {
            display(document: document, with: sourceImage)
        }
    }
}

extension GenericDocumentViewController: UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.presentingViewController?.dismiss(animated: true, completion: nil)
    }
    
    func imagePickerController(_ picker: UIImagePickerController,
                               didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        let image = info[UIImagePickerController.InfoKey.originalImage] as? UIImage
        dismiss(animated: true) { [weak self] in
            guard let self = self else { return }
            let recognizer = SBSDKGenericDocumentRecognizer(acceptedDocumentTypes: self.documentTypes())
            recognizer.accumulatesDocuments = false
            
            if let image = image, let document = recognizer.recognizeDocument(on: image)?.document {
                self.display(document: document, with: image)
            }
        }
    }
}